Source/WebKit/UIProcess/API/glib/WebKitFeature.h.in

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (C) 2023 Igalia S.L.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public License
+ * along with this library; see the file COPYING.LIB.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301, USA.
+ */
+
+@API_SINGLE_HEADER_GUARD@
+
+#ifndef WebKitFeature_h
+#define WebKitFeature_h
+
+#include <glib-object.h>
+#include <@API_INCLUDE_PREFIX@/WebKitDefines.h>
+
+G_BEGIN_DECLS
+
+/**
+ * WebKitFeatureStatus:
+ * @WEBKIT_FEATURE_STATUS_EMBEDDER:
+ * @WEBKIT_FEATURE_STATUS_UNSTABLE:
+ * @WEBKIT_FEATURE_STATUS_INTERNAL:
+ * @WEBKIT_FEATURE_STATUS_DEVELOPER:
+ * @WEBKIT_FEATURE_STATUS_TESTABLE:
+ * @WEBKIT_FEATURE_STATUS_PREVIEW:
+ * @WEBKIT_FEATURE_STATUS_STABLE:
+ * @WEBKIT_FEATURE_STATUS_MATURE:
+ *
+ * Since: 2.42
+ */
+typedef enum {
+    WEBKIT_FEATURE_STATUS_EMBEDDER,
+    WEBKIT_FEATURE_STATUS_UNSTABLE,
+    WEBKIT_FEATURE_STATUS_INTERNAL,
+    WEBKIT_FEATURE_STATUS_DEVELOPER,
+    WEBKIT_FEATURE_STATUS_TESTABLE,
+    WEBKIT_FEATURE_STATUS_PREVIEW,
+    WEBKIT_FEATURE_STATUS_STABLE,
+    WEBKIT_FEATURE_STATUS_MATURE,
+} WebKitFeatureStatus;
+
+/**
+ * WebKitFeatureCategory:
+ * @WEBKIT_FEATURE_CATEGORY_NONE:
+ * @WEBKIT_FEATURE_CATEGORY_ANIMATION:
+ * @WEBKIT_FEATURE_CATEGORY_CSS:
+ * @WEBKIT_FEATURE_CATEGORY_DOM:
+ * @WEBKIT_FEATURE_CATEGORY_HTML:
+ * @WEBKIT_FEATURE_CATEGORY_JAVASCRIPT:
+ * @WEBKIT_FEATURE_CATEGORY_MEDIA:
+ * @WEBKIT_FEATURE_CATEGORY_NETWORK:
+ * @WEBKIT_FEATURE_CATEGORY_PRIVACY:
+ * @WEBKIT_FEATURE_CATEGORY_SECURITY:
+ *
+ * Since: 2.42
+ */
+typedef enum {
+    WEBKIT_FEATURE_CATEGORY_NONE,
+    WEBKIT_FEATURE_CATEGORY_ANIMATION,
+    WEBKIT_FEATURE_CATEGORY_CSS,
+    WEBKIT_FEATURE_CATEGORY_DOM,
+    WEBKIT_FEATURE_CATEGORY_HTML,
+    WEBKIT_FEATURE_CATEGORY_JAVASCRIPT,
+    WEBKIT_FEATURE_CATEGORY_MEDIA,
+    WEBKIT_FEATURE_CATEGORY_NETWORK,
+    WEBKIT_FEATURE_CATEGORY_PRIVACY,
+    WEBKIT_FEATURE_CATEGORY_SECURITY,
+} WebKitFeatureCategory;
```

Comment:
I'm nervous about these two because I expect feature status and feature category enums to change in the future. Understanding that exposing them makes the API more powerful, I'm not sure it's worth the risk, unless there is corresponding public Cocoa API. You can do most of what you want to do without these, right? The features list just would not be sorted so nicely?

Reply:
Private/unstable is the thing... maybe find the relevant Apple developer and talk to them before making this public API?

> The feature status is quite informative and WebKit itself uses it internally to decide some defaults, and what should be considered “experimental and can be shown in the Safari TP ‘Develop’ menu”; or otherwise “definitely internal or very much in development and applications shouldn't show this”. So at least I would like to keep WebKitFeatureStatus.

There are so many status types... really a lot. I would fear these could be consolidated in the future. Realistically, there are so many states and not amazing distinctions between them (e.g. "stable" vs. "mature") that

We can at least try to simplify somewhat at the public layer. I'd say UNSTABLE, DEVELOPER, TESTABLE, and PREVIEW can all be compressed together. STABLE and MATURE can be compressed. That leaves EMBEDDER (not quite sure about the intended semantics of this) and INTERNAL (what is this? I guess we should just hide features that are INTERNAL?). The fewer statuses we expose, the lower the probability that we have trouble and have to deprecate statuses in the future.  

> If we are wary about the values of the WebKitFeatureCategory changing between releases, we could make the category a string, and then we would be free to return different strings as releases go by. It would be a bit annoying to not know in advance how many categories are there before enumerating the features, but it would be more flexible in exchange for making the API a bit less convenient. Thoughts?

Problem solved. I was going to suggest exactly that.

Reply:
> PREVIEW

I suppose this status could allow us to enable certain features in Tech Preview while leaving them disabled elsewhere, so maybe it could be useful... but we can decide that later. It's easy to change our mind and expose a new category in the future. But we can't get rid of a status once we expose it; the best we can do is deprecate and just not have any features associated with that status anymore.

Reply:
Please check the updated patch, which has most of the documentation and includes descriptions for all the `WebKitFeatureStatus` enum values, and let me know if you still think they are unclear and/or some should be merged. To me all statuses look useful.

And indeed, we can use the status to decide which features to enable by default in Ephy TP, and which ones to show in some UI there to allow toggling. Note that `webkit_settings_get_{experimental,development}_features()` are convenience functions that return feature lists filtered by the status 😉. So what Ephy TP most likely would do is pick the experimental list, and enable “preview” features by default, and allow toggling all of “testable”, “preview”, and “stable” (maybe even “developer”).